File: Tutorials/OpenGLLayerExample/netcore/MainWindow.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace OpenGLLayerExample
{
   public class MainWindow : ActiveWindow
   {
      public override void InitWindow()
      {
         SetParameters(this.GetType().Name, this.GetType().Name, 800, 800, false);
         IsCentered = true;

         OneCubeExample();
         // MultipleCubes();
      }

      private void OneCubeExample()
      {
         TitleBar titleBar = new TitleBar(this.GetType().Name);

         OpenGLLayer ogl = new OpenGLLayer();
         ogl.SetMargin(0, titleBar.GetHeight(), 0, 0);

         HorizontalStack toolbar = Items.GetToolbarLayout();

         ImagedButton btnRotateLeft = Items.GetImagedButton(EmbeddedImage.ArrowUp, -90);
         ImagedButton btnRotateRight = Items.GetImagedButton(EmbeddedImage.ArrowUp, 90);

         HorizontalSlider zoom = Items.GetSlider();

         ImagedButton btnRestoreView = Items.GetImagedButton(EmbeddedImage.Refresh, 0);

         // adding
         AddItems(titleBar, ogl);
         ogl.AddItems(toolbar);
         toolbar.AddItems(btnRotateLeft, btnRotateRight, zoom, btnRestoreView);

         // assign events
         btnRestoreView.EventMousePress += (sender, args) =>
         {
            ogl.RestoreView();
         };

         btnRotateLeft.EventMousePress += (sender, args) =>
         {
            ogl.Rotate(KeyCode.Left);
         };

         btnRotateRight.EventMousePress += (sender, args) =>
         {
            ogl.Rotate(KeyCode.Right);
         };

         zoom.EventValueChanged += (sender) =>
         {
            ogl.SetZoom(zoom.GetCurrentValue());
         };

         // Set focus
         ogl.SetFocus();
         zoom.SetCurrentValue(3);
      }

      private void MultipleCubes()
      {
         TitleBar titleBar = new TitleBar(this.GetType().Name);

         FreeArea area = new FreeArea();
         area.SetMargin(0, titleBar.GetHeight(), 0, 0);

         AddItems(titleBar, area);

         List<IBaseItem> content = new List<IBaseItem>();

         for (int row = 0; row < 3; row++)
         {
            for (int column = 0; column < 3; column++)
            {
               ResizableItem frame = new ResizableItem();
               frame.SetBorder(new Border(Color.Gray, new CornerRadius(), 2));
               frame.SetPadding(5, 5, 5, 5);
               frame.SetBackground(100, 100, 100);
               frame.SetSize(200, 200);
               frame.SetPosition(90 + row * 210, 60 + column * 210);
               area.AddItem(frame);
               content.Add(frame);

               frame.EventMousePress += (sender, args) =>
               {
                  content.Remove(frame);
                  content.Add(frame);
                  area.SetContent(content);
               };

               OpenGLLayer ogl = new OpenGLLayer();
               ogl.SetMargin(0, 30, 0, 0);
               frame.AddItem(ogl);
            }
         }
      }
   }
}

File: Tutorials/OpenGLLayerExample/java/src/main/java/OpenGLLayerExample/MainWindow.java
package OpenGLLayerExample;

import java.util.*;
import java.awt.Color;

import com.spvessel.spacevil.*;
import com.spvessel.spacevil.Core.*;
import com.spvessel.spacevil.Flags.*;
import com.spvessel.spacevil.Decorations.*;

public class MainWindow extends ActiveWindow {

    @Override
    public void initWindow() {
        setParameters(this.getClass().getName(), this.getClass().getName(), 800, 800, false);
        isCentered = true;

        // oneCubeExample();
        multipleCubes();
    }

    private void oneCubeExample() {
        TitleBar titleBar = new TitleBar(this.getClass().getName());

        OpenGLLayer ogl = new OpenGLLayer();
        ogl.setMargin(0, titleBar.getHeight(), 0, 0);

        HorizontalStack toolbar = Items.getToolbarLayout();

        ImagedButton btnRotateLeft = Items.getImagedButton(EmbeddedImage.ARROW_UP, -90);
        ImagedButton btnRotateRight = Items.getImagedButton(EmbeddedImage.ARROW_UP, 90);

        HorizontalSlider zoom = Items.getSlider();

        ImagedButton btnRestoreView = Items.getImagedButton(EmbeddedImage.REFRESH, 0);

        // adding
        addItems(titleBar, ogl);
        ogl.addItems(toolbar);
        toolbar.addItems(btnRotateLeft, btnRotateRight, zoom, btnRestoreView);

        // assign events
        btnRestoreView.eventMousePress.add((sender, args) -> {
            ogl.restoreView();
        });

        btnRotateLeft.eventMousePress.add((sender, args) -> {
            ogl.rotate(KeyCode.LEFT);
        });

        btnRotateRight.eventMousePress.add((sender, args) -> {
            ogl.rotate(KeyCode.RIGHT);
        });

        zoom.eventValueChanged.add((sender) -> {
            ogl.setZoom(zoom.getCurrentValue());
        });

        // set focus
        ogl.setFocus();
        zoom.setCurrentValue(3);
    }

    private void multipleCubes() {
        TitleBar titleBar = new TitleBar(this.getClass().getName());

        FreeArea area = new FreeArea();
        area.setMargin(0, titleBar.getHeight(), 0, 0);

        addItems(titleBar, area);

        List<InterfaceBaseItem> content = new ArrayList<>();

        for (int row = 0; row < 3; row++) {
            for (int column = 0; column < 3; column++) {
                ResizableItem frame = new ResizableItem();
                frame.setBorder(new Border(Color.gray, new CornerRadius(), 2));
                frame.setPadding(5, 5, 5, 5);
                frame.setBackground(100, 100, 100);
                frame.setSize(200, 200);
                frame.setPosition(90 + row * 210, 60 + column * 210);
                area.addItem(frame);
                content.add(frame);

                frame.eventMousePress.add((sender, args) -> {
                    content.remove(frame);
                    content.add(frame);
                    area.setContent(content);
                });

                OpenGLLayer ogl = new OpenGLLayer();
                ogl.setMargin(0, 30, 0, 0);
                frame.addItem(ogl);
            }
        }
    }
}
